pkg/compiler/check.go

```diff
@@ -916,21 +920,23 @@ func (comp *compiler) replaceTypedef(ctx *checkCtx, t *ast.Type, flags checkFlag
 		}
 	}
 	fullTypeName := ast.SerializeNode(t)
+	recursion := 0
 	for i, prev := range ctx.instantiationStack {
-		if prev == fullTypeName {
-			ctx.instantiationStack = append(ctx.instantiationStack, fullTypeName)
-			path := ""
-			for j := i; j < len(ctx.instantiationStack); j++ {
-				if j != i {
-					path += " -> "
-				}
-				path += ctx.instantiationStack[j]
-			}
-			comp.error(t.Pos, "type instantiation loop: %v", path)
-			return
+		if typedefName == templateBase(prev) {
+			recursion++
```

Comment:
1) Wouldn't it more reliable to store the recursion depth in the stack?
2) Wouldn't it be better to move the `if recursion > 10 {` check here?

Reply:
> Wouldn't it be better to move the if recursion > 10 { check here?

Done.
I also restructured where we do `ctx.instantiationStack = append(ctx.instantiationStack, fullTypeName)`, otherwise we would need to duplicate this statement 3 times.

> Wouldn't it more reliable to store the recursion depth in the stack?

I don't see how to do it easily.
I calculate recursion depth per type, so we would need to add map[string]int to the context. But then we would need to make a copy of this map as we recurse and return from recursion. And making a copy of the map would require a loop. So some amount of additional code.